src/aioswitcher/api/__init__.py

```diff
@@ -103,6 +108,10 @@ def __init__(
         self._device_key = device_key
         self._port = port
         self._connected = False
+        if self._device_type.token_needed and bool(token):
```

Comment:
1. Because of this reason:
- We don't need to convert the token to a packet if the device does not require a token.. This is to prevent selecting a device that does not need a token and providing a token.
2. It should send the API calls to the token-based packets with an empty string token and will result in no action to the device.
3. Because self._token is later used for the packet sent to the device it should be at least an empty string.. I could split that into 2 variables: self._token and self._token_packet if this makes more sense, I thought it just needed..

Reply:
> It doesn't make sense to convert a type of str or None to a Boolean, why not just checking if it is not None?

I'm ok with _bool(token)_. The [bool](https://www.w3schools.com/python/ref_func_bool.asp) function in Python is used more to evaluate expressions than convert. It's commonly used and will return false for `None` and `""`.

The other two questions still stand.

Reply:
@YogevBokobza so it is not missed, just reminding that above 2 & 3 still need attention

Reply:
I thought I already answered lol
Anyway here:
Regarding (2) if a token is needed but is not provided that is what the "if" is for, it should set the token to be empty and the packet will be sent ok but won't do anything in practice.. no control would occur..
Regarding (3) I explained why it cannot stay None... we need to send some string as a token. I said I could set another variable like "self._token_packet" if we don't want to use the same one..

Reply:
> packet will be sent ok but won't do anything in practice.. no control would occur.

So what is the point of even checking that? You detected that a token is needed, but it was not provided and you silently ignore it let it just don't work at a later stage. If a token is needed and not provided raise an error, this will also remove the need to constantly check this in later calls.

> explained why it cannot stay None... we need to send some string as a token

This again gets back to (2), if you repeatedly check if a token is set:
https://github.com/TomerFi/aioswitcher/blob/84bc398ec5b55ceb3f7c7e082e25091e207a4409/src/aioswitcher/api/__init__.py#L174
https://github.com/TomerFi/aioswitcher/blob/84bc398ec5b55ceb3f7c7e082e25091e207a4409/src/aioswitcher/api/__init__.py#L193
https://github.com/TomerFi/aioswitcher/blob/84bc398ec5b55ceb3f7c7e082e25091e207a4409/src/aioswitcher/api/__init__.py#L833
https://github.com/TomerFi/aioswitcher/blob/84bc398ec5b55ceb3f7c7e082e25091e207a4409/src/aioswitcher/api/__init__.py#L945

If you evaluate if a token is needed correctly at (2) you will never reach to a situation that you send an empty token, so I would expect the init flow to be:
```python
        self._token = ""
        if self._device_type.token_needed:
            if not token:
                raise RuntimeError("A token is needed but is missing")
            self._token = convert_token_to_packet(str(token))
```
So all of this places can just change from:
`if self._device_type and self._device_type.token_needed and bool(self._token):`
to:
`if self._token:`

Reply:
All of those `if self._device_type and self._device_type.token_needed and bool(self._token): ` lines are gone in the past commits.. please see the latest changes.
I replaced it with the code suggested and added a test to support this Error Raise part.